### DISCUSSION TITLE
Is there an equivalent of `getServerSideProps` from Next.js?

### DISCUSSION THREAD

**Discussion by muuvmuuv** [gatsbyjs/gatsby#29655] [Help]
I would like to pre-render/change a prop I pass to a component before it is delivered to the browser. With `getServerSideProp` I could do something like that I think. Does Gatsby have any API/Function to change properties passed to a component on build-time/SSR?

Something like:

```jsx
import React from 'react'

export const onServerSideRendering = ({ something }) => {
  something = "somethingelse"
}

const Test = ({ something, ...props }) => {
  return (
    <p {...props}>
      {something} {/* => somethingelse */}
    </p>
  )
}

export { Test }

```

**Comment by nibtime** [comment]
@muuvmuuv  

[`getServerSideProps`](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering) of Next pulls props into a specific page at request time before it is served to a visitor. Gatsby is a SSG only, not a hybrid Framework like Next (and for SSG and SSR pages side-by-side you also have to host on Vercel), so I really have no idea how you could do something like that with Gatsby at all. 

Do you mean [`getStaticProps`](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation), which allows you to pull props into a specific page during SSG by a function of type `() => Promise<T>`?  I am missing that sorely from Gatsby, that's why I am trying to write a theme that emulates `getStaticProps`. I opened up a [discussion ](https://github.com/gatsbyjs/gatsby/discussions/29622) for that. If that's the case I would be interested in your specific use case be glad if you'd share it.

**Reply by muuvmuuv** [reply]
Yes get static props would be the same in that case I guess.

I am writing a E-Mail protecting component "Mailto" which hashes the mail in the mailto link or the content itself it it is an email. That needs to happen before the page is delivered to the user because crawlers typically disable JavaScript to avoid such hacking. Users with disabled JavaScript will just see a "mailto:?Subject=Please enable JavaScript" and with enabled JavaScript they can click on the link and the mailto will be decoded and opened through JS. That way I think crawlers can be tricked. I dont think they can "catch" an `window.location.href=""`.

**Reply by nibtime** [reply]
Do you want this hashing/server-side rendering to happen on a per-request basis (`getServerSideProps`) or a per-build basis (`getStaticProps`)? Can you maybe share a little code snippet of the component that can help me understand it a little bit better?

In `getServerSideProps` you can make use of additional parameters to take influence on the control flow of the Server-Side Rendering, i.e. the Node HTTP incomingmessage and response object + Next adds some convenience params, like offering you the query parameters as an object.

In Next, per page, you have to use either `getStaticProps` (to say it should be server-rendered to HTML once at build-time - SSG) or `getServerSideProps` (to say it should be server-rendered to HTML on a per-request basis -SSR). With the latter, you could do things like passing client-side data from user input of the "Mailto" component as query parameters when navigating to the page and then do the hashing on the server and serve pre-rendered HTML based on that.

In Next there's even a 3rd option (incremental static generation - ISG), which is integrated with the Vercel Edge Network. This allows you to say something like "repeat the static generation of the page every 60s but also only upon the first request of a new visitor" (so it's lazy). This is for use cases where the data of your page changes very often while still retaining most of the benefits of static sites (for situations where you would otherwise do client-side-rendering (CSR) upon data fetched from a database or API, or if you have couple millions of pages). However this works with `getStaticProps`, so you can't prerender HTML dependent on request parameters like you can with `getServerSideProps` here either.

If you find, that for the site(s) you are making you would frequently need `getServerSideProps` or incremental static generation, I would really recommend you to give Next a try, as this is not and probably never will be possible with Gatsby. I am sort of torn myself at the moment between the two (Next 10 added some huge things for me), but don't want to drop all my Gatsby knowledge either. But "just goddamn migrate" is always prevalent in the back of my head while I am authoring my current theme 😄

**Reply by muuvmuuv** [reply]
Thanks for that detailed explanation! In our case `per-build basis` because we don't have any dynamic "email content". I am in the exact same situation. Build several sites with Gatsby but after building just one with Next I feel its more future prove because of its massive amount of features and the support it becomes. Only thing I missed in Next was sharp intergration to produce optimized image assets AVIF → webp → JPEG2000 + automated picture generation with different resolutions and width/height inlining to reduce layout shifts.

**Reply by ascorbic** [reply]
The simplest way to do this is to update it inside a useEffect hook. That would not be rendered in SSR, buit only after hydration. So something like:

```jsx

export function Mailto({ address, ...props }) {
  const [toAddress, setToAddress] = useState("mailto:?Subject=Please enable JavaScript")
  useEffect(() => {
   setToAddress(`mailto:${address}`)
  }, [address])

  return <a href={address} {...props} />
}
```

**Reply by muuvmuuv** [reply]
Wouldn't the email still be visible in the compiled code? The Point is, crawlers are tricky these days and probably search through downloaded scripts as well.

PS: `toAddress` → `address` ^^

**Reply by ascorbic** [reply]
You could do whatever scrambling you wanted on the prop, and unscramble it in the hook

**Reply by nibtime** [reply]
@muuvmuuv you're welcome 😃. I am glad if it can assist you in making a good decision for yourself/your team. It's really great to have options like Gatsby and Next available. Image optimization with sharp is also one of the most important parts of Gatsby for me. Next 10 added first-class support for image optimization, however, it's dynamic and I didn't try it out yet. There's also a new [`gatsby-plugin-image`](https://www.npmjs.com/package/gatsby-plugin-image) with a, I find, much-improved interface and with convenient support for local and remote images through a `<StaticImage>` component. I will certainly compare them at some point to see which works better for me.

**Comment by nibtime** [comment]
@muuvmuuv 
Maybe the interception pattern you are looking for can be implemented with a wrapper component that makes use of `useStaticQuery`? The following would yield a component that just behaves like your Mail component, but swaps certain props with static values during SSR:

<details>
<summary>create node field with hash (for useStaticQuery later)</summary>

```js
const { createHash } = require("crypto")

exports.onCreateNode = ({node, actions}) => {
  if(node?.frontmatter?.mailto) {
    const hash = createHash("sha256")
    hash.update(node.frontmatter.mailto)
    const digest = hash.digest().toString("hex")
    actions.createNodeField({
      name: "mailtoHash",
      node,
      value: digest
    })
  }

}
```

</details>


<details>
<summary>wrapper component for "static interception"</summary>

```tsx
import React from 'react'
import { graphql } from 'gatsby'

const MailSSRIntercepted = ({ mailto, ...props }) => {
  const data = useStaticQuery(graphql`
    mdx(...) {
      fields {
        mailtoHash
      }
      frontmatter {
        mailto
      }
    }
  `)

  return <Mail mailto={typeof window === 'undefined' ? `${data.mdx.frontmatter.mailto}${data.mdx.fields.mailtoHash}` : mailto} {...props } />
}
```

</details>

**Reply by muuvmuuv** [reply]
That a good idea! Will try it out. We just moved that to a later sprint to focus on other things for now.
